src/cellfinder_core/detect/filters/volume/structure_detection.py

```diff
@@ -238,89 +255,78 @@ def connect_eight(self, layer):
-
     def get_coords_dict(self):
         return self.coords_maps
 
-    def add(self, x: int, y: int, z: int, neighbour_ids):
+    def add_point(self, sid: int, point: np.ndarray) -> None:
+        """
+        Add *point* to the structure with the given *sid*.
+        """
+        self.coords_maps[sid] = np.row_stack((self.coords_maps[sid], point))
+
+    def add(self, x: int, y: int, z: int, neighbour_ids: List[int]) -> int:
         """
         For the current coordinates takes all the neighbours and find the
-        minimum structure
-        including obsolete structures mapping to any of the neighbours
-        recursively.
+        minimum structure including obsolete structures mapping to any of
+        the neighbours recursively.
+
         Once the correct structure id is found, append a point with the
-        current coordinates to the coordinates map
-        entry for the correct structure. Hence each entry of the map will be a
-        vector of all the pertaining points.
-
-        :param x:
-        :param y:
-        :param z:
-        :param neighbour_ids:
-        :return:
+        current coordinates to the coordinates map entry for the correct
+        structure. Hence each entry of the map will be a vector of all the
+        pertaining points.
         """
         updated_id = self.sanitise_ids(neighbour_ids)
+        if updated_id not in self.coords_maps:
+            self.coords_maps[updated_id] = np.zeros(
+                shape=(0, 3), dtype=np.uint64
+            )
         self.merge_structures(updated_id, neighbour_ids)
 
-        p = Point(x, y, z)  # Necessary to split definition on some machines
-        self.coords_maps[updated_id].append(p)  # Add point for that structure
-
+        # Add point for that structure
+        point = np.array([[x, y, z]], dtype=np.uint64)
+        self.add_point(updated_id, point)
         return updated_id
 
-    def sanitise_ids(self, neighbour_ids):
+    def sanitise_ids(self, neighbour_ids: List[int]) -> int:
         """
+        Get the smallest ID of all the structures that are connected to IDs
+        in `neighbour_ids`.
+
         For all the neighbour ids, walk up the chain of obsolescence (self.
-        obsolete_ids)
-        to reassign the corresponding most obsolete structure to the current
-        neighbour
+        obsolete_ids) to reassign the corresponding most obsolete structure
+        to the current neighbour.
 
-        :param neighbour_ids:
-        :return: updated_id
+        Has no side effects on this class.
         """
-        for i in range(len(neighbour_ids)):
-            neighbour_id = neighbour_ids[i]
+        for i, neighbour_id in enumerate(neighbour_ids):
             # walk up the chain of obsolescence
-            while neighbour_id in self.obsolete_ids:
-                neighbour_id = self.obsolete_ids[neighbour_id]
+            neighbour_id = int(traverse_dict(self.obsolete_ids, neighbour_id))
```

Comment:
nice!

Reply:
This was actually to work around a `numba` peculiarity and I was unsure if haivng a funciton here was clearer, but glad you think it is 😄